plonky2/src/hash/poseidon2_bn254.rs

```diff
@@ -1,149 +1,323 @@
+
+        perm.permute();
+        let out = perm.squeeze();
+
+        BN254HashOut {
+            element: out[0].clone(),
+        }
+    }
+
+    fn squeeze_goldilocks(state: &mut Self::Permutation) -> Vec<F> {
+        let bn_out = state.squeeze();
+        let bn_bytes: Vec<u8> = bn_out.iter().flat_map(|e| felts_to_bytes(e)).collect();
+        let goldilocks_felts: Vec<F> = bytes_to_u64(&bn_bytes).iter().map(|e| F::from_canonical_u64(*e)).collect();
+        assert!(goldilocks_felts.len()>0);
+        goldilocks_felts
+    }
+}
+
+// --------- Conversion helper functions ---------------------
+
+/// converts a vec of goldilocks to bn254
+/// takes 7 goldilocks and converts to 2 bn254
+fn goldilocks_to_bn<F: RichField>(input: &Vec<F>) -> Vec<BN254Fr>{
+    let u64s: Vec<u64> = input.iter().map(|x| x.to_canonical_u64()).collect();
+    let l = u64s.len();
+    let m = l / 7;
+    let mut result = Vec::new();
+
+    for i in 0..m {
+        let group: [u64; 7] = u64s[7 * i..7 * (i + 1)].try_into().unwrap();
+        let (a, b) = u64s_to_felts(group);
+        result.push(a);
+        result.push(b);
     }
+
+    let r = l - 7 * m;
+    if r > 0 {
+        let mut ws = [0u64; 7];
+        for i in 0..r {
+            ws[i] = u64s[7 * m + i];
+        }
+        let (a, b) = u64s_to_felts(ws);
+        // check that we don't push zero field elements
+        if a != BN254Fr::zero() {
+            result.push(a);
+        }
+        if b != BN254Fr::zero() {
+            result.push(b);
+        }
+    }
+    result
+}
+
+const BIGINT_TWO_TO_64:  BigInt<4> = BigInt( [0,1,0,0] );
+const BIGINT_TWO_TO_128: BigInt<4> = BigInt( [0,0,1,0] );
+const BIGINT_TWO_TO_192: BigInt<4> = BigInt( [0,0,0,1] );
+
+/// converts u64 to BN254 - taken directly from: rust-bn254-hash
+pub fn u64s_to_felts(ws: [u64; 7]) -> (BN254Fr, BN254Fr) {
+    let hi = ws[6] >> 32;
+    let lo = ws[6] & 0xFFFF_FFFF;
+
+    let field_powers_of_two_to_64: [BN254Fr;3] =
+        [
+            BN254Fr::from_bigint(BIGINT_TWO_TO_64 ).unwrap(),
+            BN254Fr::from_bigint(BIGINT_TWO_TO_128).unwrap(),
+            BN254Fr::from_bigint(BIGINT_TWO_TO_192).unwrap()
+        ];
+
+    let x = BN254Fr::from(ws[0])
+        + field_powers_of_two_to_64[0] * BN254Fr::from(ws[1])
+        + field_powers_of_two_to_64[1] * BN254Fr::from(ws[2])
+        + field_powers_of_two_to_64[2] * BN254Fr::from(lo);
+
+    let y = BN254Fr::from(ws[3])
+        + field_powers_of_two_to_64[0] * BN254Fr::from(ws[4])
+        + field_powers_of_two_to_64[1] * BN254Fr::from(ws[5])
+        + field_powers_of_two_to_64[2] * BN254Fr::from(hi);
+
+    (x, y)
 }
 
-fn felts_to_bytes<E>(f: E) -> Vec<u8> where
+/// converts a slice of bytes to 64 by taking 63 bits at a time
+/// this makes it safe for conversion from bytes to Goldilocks field elems
+/// this fn ignores any remaining bit that are less than 63 bits at the end
+pub fn bytes_to_u64(x: &[u8]) -> Vec<u64> {
```

Comment:
my understanding is that this always outputs elements where the most significant bit is zero (that is, they are `0 <= x < 2^63`). That's not a good idea...